enterprise/backend/test/metabase_enterprise/serialization/cmd_test.clj

```diff
@@ -77,4 +87,117 @@
+     (do-with-in-memory-h2-db
+      "dest-"
+      (fn [do-with-dest-db]
+        (f do-with-source-db do-with-dest-db))))))
+
+(defmacro ^:private with-source-and-dest-dbs {:style/indent 0} [& body]
+  `(do-with-source-and-dest-dbs
+    (fn [~'&do-with-source-db ~'&do-with-dest-db]
+      ~@body)))
+
+(defmacro ^:private with-source-db {:style/indent 0} [& body]
+  `(~'&do-with-source-db (fn [] ~@body)))
+
+(defmacro ^:private with-dest-db {:style/indent 0} [& body]
+  `(~'&do-with-dest-db (fn [] ~@body)))
+
+(defn- do-with-random-dump-dir [f]
+  (let [dump-dir (random-dump-dir)]
+    (testing (format "\nDump dir = %s" (pr-str dump-dir))
+      (try
+        (f dump-dir)
+        (finally
+          (when (.exists (io/file dump-dir))
+            (.delete (io/file dump-dir))))))))
+
+(defmacro ^:private with-random-dump-dir {:style/indent 1} [[dump-dir-binding] & body]
+  `(do-with-random-dump-dir (fn [~dump-dir-binding] ~@body)))
+
+(deftest mode-update-remove-cards-test
+  (testing "--mode update should remove Cards in a Dashboard if they're gone from the serialized YAML (#20786)"
+    (with-random-dump-dir [dump-dir]
+      (let [dashboard-yaml-filename (str dump-dir "/collections/root/dashboards/Dashboard.yaml")]
+        (with-source-and-dest-dbs
+          (testing "create 2 questions in the source and add them to a dashboard"
+            (with-source-db
+              (let [{db-id :id, :as db} (create! Database :name "My_Database")]
+                (mt/with-db db
+                  (let [{user-id :id}      (create! User, :is_superuser true)
+                        {card-1-id :id}    (create! Card
+                                                    :database_id   db-id
+                                                    :creator_id    user-id
+                                                    :name          "Card_1"
+                                                    :dataset_query {:database db-id, :type :native, :native {:query "SELECT 1;"}})
+                        {card-2-id :id}    (create! Card
+                                                    :database_id   db-id
+                                                    :creator_id    user-id
+                                                    :name          "Card_2"
+                                                    :dataset_query {:database db-id, :type :native, :native {:query "SELECT 1;"}})
+                        {dashboard-id :id} (create! Dashboard, :creator_id user-id, :name "Dashboard")]
+                    (doseq [card-id [card-1-id card-2-id]]
+                      (create! DashboardCard :dashboard_id dashboard-id, :card_id card-id))
+                    (testing "dump in source"
+                      (is (nil? (cmd/dump dump-dir)))))))))
+          (testing "verify the Dashboard was dumped as expected"
+            (is (.exists (io/file dashboard-yaml-filename)))
+            (when-let [yaml (yaml/from-file dashboard-yaml-filename)]
```

Comment:
maybe change the `when-let` to an `if-let` who's else branch is a test failure in case of bad yaml? or is that overkill since the file exists? If the file were empty does `yaml/form-file` throw or return nil?

Reply:
`touch empty.yaml` and then `(yaml/from-file "empty.yaml")` will return nil here.